util/schema.go

```diff
@@ -0,0 +1,326 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/openconfig/goyang/pkg/yang"
+)
+
+// IsLeafRef reports whether schema is a leafref schema node type.
+func IsLeafRef(schema *yang.Entry) bool {
+	if schema == nil || schema.Type == nil {
+		return false
+	}
+	return schema.Type.Kind == yang.Yleafref
+}
+
+// IsChoiceOrCase returns true if the entry is either a 'case' or a 'choice'
+// node within the schema. These are schema nodes only, and the code generation
+// operates on data tree paths.
+func IsChoiceOrCase(e *yang.Entry) bool {
+	if e == nil {
+		return false
+	}
+	return e.IsChoice() || e.IsCase()
+}
+
+// SchemaPaths returns all the paths in the path tag.
+func SchemaPaths(f reflect.StructField) ([][]string, error) {
```

Comment:
We have multiple implementations of this function. I wonder whether we can make them common. We should create an issue.

Reply:
https://github.com/openconfig/ygot/issues/120